File: typescript/src/verification/invariant/structural-check.ts
/**
 * @module structural-check
 *
 * Structural deadlock pre-check using siphon/trap analysis (Commoner's theorem).
 *
 * **Commoner's theorem**: A Petri net is deadlock-free if every siphon contains
 * an initially marked trap.
 *
 * **Siphon**: A set of places S where every transition that outputs to S also
 * inputs from S. Key property: once all places in a siphon become empty,
 * they can never be re-marked. An empty siphon can cause deadlock.
 *
 * **Trap**: A set of places S where every transition that inputs from S also
 * outputs to S. Key property: once any place in a trap is marked,
 * the trap remains marked forever.
 *
 * **Algorithm**: Find every minimal siphon by a branching search: grow a set from
 * each place and, where a producer into the set has no input in it, branch on each
 * of that producer's inputs. For each minimal siphon, find its maximal trap
 * (fixed-point contraction). If every such trap is marked in the initial marking,
 * deadlock-freedom is proven structurally and no SMT query is needed. The search is
 * capped at SIPHON_SEARCH_BUDGET nodes; past it the check is inconclusive.
 *
 * Limited to nets with ≤50 places to bound enumeration cost.
 */
import type { FlatNet } from '../encoding/flat-net.js';
import type { MarkingState } from '../marking-state.js';

const MAX_PLACES_FOR_SIPHON_ANALYSIS = 50;

/**
 * Search budget for the siphon search, in search nodes. Deciding Commoner's
 * condition is co-NP-complete, so the search is exponential in the worst case;
 * past this budget the check answers `inconclusive` and the SMT pipeline
 * decides. Identical in every implementation.
 */
const SIPHON_SEARCH_BUDGET = 10_000;

/**
 * Result of structural deadlock check using siphon/trap analysis.
 */
export type StructuralCheckResult =
  | { readonly type: 'no-potential-deadlock' }
  | { readonly type: 'potential-deadlock'; readonly siphon: ReadonlySet<number> }
  | { readonly type: 'inconclusive'; readonly reason: string };

/**
 * Structural deadlock pre-check using siphon/trap analysis.
 *
 * Commoner's theorem: a Petri net is deadlock-free if every siphon
 * contains a marked trap.
 *
 * A siphon is a set of places S such that every transition with
 * an output in S also has an input in S. Once empty, a siphon stays empty.
 *
 * A trap is a set of places S such that every transition with
 * an input in S also has an output in S. Once marked, a trap stays marked.
 *
 * Checking every *minimal* siphon suffices, since a trap inside a minimal siphon
 * lies inside every siphon containing it. So `no-potential-deadlock` needs the
 * search to have found **every** minimal siphon, and each one's maximal trap to
 * hold a token in the initial marking.
 */
export function structuralCheck(flatNet: FlatNet, initialMarking: MarkingState): StructuralCheckResult {
  const P = flatNet.places.length;

  if (P === 0) {
    return { type: 'no-potential-deadlock' };
  }

  if (P > MAX_PLACES_FOR_SIPHON_ANALYSIS) {
    return { type: 'inconclusive', reason: `Net has ${P} places, siphon enumeration skipped` };
  }

  const siphons = findMinimalSiphons(flatNet, SIPHON_SEARCH_BUDGET);
  if (siphons === null) {
    return { type: 'inconclusive', reason: `siphon search exceeded ${SIPHON_SEARCH_BUDGET} nodes` };
  }

  for (const siphon of siphons) {
    const trap = findMaximalTrapIn(flatNet, siphon);

    if (trap.size === 0 || !isMarked(trap, flatNet, initialMarking)) {
      return { type: 'potential-deadlock', siphon };
    }
  }

  return { type: 'no-potential-deadlock' };
}

/**
 * Finds all minimal siphons (indices into `flatNet.places`). Exponential in the
 * worst case. Pass `budget` to cap the search; past `budget` search nodes it
 * returns `null`.
 *
 * A siphon is a place set `S` such that every transition with an output in `S`
 * has at least one input in `S`. The search grows `S` from each start place;
 * where a producer into `S` has no input in `S`, it branches on each of that
 * producer's inputs. Every minimal siphon containing the start place is reached
 * by the branch that always picks an input inside it, so the search is complete.
 * Committing to one input instead (the first, or all of them at once) is not:
 * it can miss exactly the unmarked siphon that makes the net dead.
 */
export function findMinimalSiphons(flatNet: FlatNet): ReadonlySet<number>[];
export function findMinimalSiphons(flatNet: FlatNet, budget: number): ReadonlySet<number>[] | null;
export function findMinimalSiphons(
  flatNet: FlatNet, budget = Number.POSITIVE_INFINITY,
): ReadonlySet<number>[] | null {
  const P = flatNet.places.length;
  const found: Set<number>[] = [];
  let nodes = 0;

  const grow = (siphon: Set<number>): boolean => {
    if (++nodes > budget) return false;
    // A superset of a siphon already found cannot lead to a new minimal one.
    if (found.some(f => isSubsetOf(f, siphon))) return true;

    const violating = flatNet.transitions.find(ft => {
      let outputsToSiphon = false;
      let hasInputInSiphon = false;
      for (const p of siphon) {
        if (ft.postVector[p]! > 0) outputsToSiphon = true;
        if (ft.preVector[p]! > 0) hasInputInSiphon = true;
      }
      return outputsToSiphon && !hasInputInSiphon;
    });

    if (violating === undefined) {
      found.push(siphon);
      return true;
    }
    // A producer with no inputs keeps any set holding its output marked: no siphon
    // on this branch. Otherwise branch on each input.
    for (let q = 0; q < P; q++) {
      if (violating.preVector[q]! > 0 && !grow(new Set([...siphon, q]))) return false;
    }
    return true;
  };

  for (let start = 0; start < P; start++) {
    if (!grow(new Set([start]))) return null;
  }

  return found.filter((s, i) => !found.some((other, j) => j !== i && isSubsetOf(other, s)));
}

/**
 * Finds the maximal trap within a given set of places.
 * Uses fixed-point: start with the full set and remove places that violate the trap condition.
 */
export function findMaximalTrapIn(flatNet: FlatNet, places: ReadonlySet<number>): ReadonlySet<number> {
  const trap = new Set(places);

  let changed = true;
  while (changed) {
    changed = false;
    const toRemove: number[] = [];

    for (const p of trap) {
      let satisfies = true;
      for (let t = 0; t < flatNet.transitions.length; t++) {
        const ft = flatNet.transitions[t]!;
        if (ft.preVector[p]! > 0) {
          let outputsToTrap = false;
          for (const q of trap) {
            if (ft.postVector[q]! > 0) {
              outputsToTrap = true;
              break;
            }
          }
          if (!outputsToTrap) {
            satisfies = false;
            break;
          }
        }
      }
      if (!satisfies) {
        toRemove.push(p);
      }
    }

    if (toRemove.length > 0) {
      for (const p of toRemove) trap.delete(p);
      changed = true;
    }
  }

  return trap;
}

function isMarked(placeIndices: ReadonlySet<number>, flatNet: FlatNet, marking: MarkingState): boolean {
  for (const idx of placeIndices) {
    const place = flatNet.places[idx]!;
    if (marking.tokens(place) > 0) return true;
  }
  return false;
}

function isSubsetOf(sub: ReadonlySet<number>, sup: ReadonlySet<number>): boolean {
  if (sub.size > sup.size) return false;
  for (const v of sub) {
    if (!sup.has(v)) return false;
  }
  return true;
}

File: java/src/main/java/org/libpetri/smt/invariant/StructuralCheck.java
package org.libpetri.smt.invariant;

import org.libpetri.analysis.MarkingState;
import org.libpetri.smt.encoding.FlatNet;
import org.libpetri.smt.encoding.FlatTransition;

import java.util.*;

/**
 * Structural deadlock pre-check using siphon/trap analysis.
 *
 * <p>Commoner's theorem: a Petri net is deadlock-free if every siphon
 * contains a marked trap.
 *
 * <p>A <b>siphon</b> is a set of places S such that every transition with
 * an output in S also has an input in S. Once empty, a siphon stays empty.
 *
 * <p>A <b>trap</b> is a set of places S such that every transition with
 * an input in S also has an output in S. Once marked, a trap stays marked.
 *
 * <p>Checking every <em>minimal</em> siphon suffices, since a trap inside a minimal
 * siphon lies inside every siphon containing it. So {@code NoPotentialDeadlock} needs
 * the search to have found <b>every</b> minimal siphon, and each one's maximal trap to
 * hold a token in the initial marking. Deciding that is co-NP-complete, so the siphon
 * search runs under a budget and answers {@code Inconclusive} past it, leaving the
 * verdict to the SMT analysis.
 */
public final class StructuralCheck {

    private static final int MAX_PLACES_FOR_SIPHON_ANALYSIS = 50;

    /** Search budget for the siphon search, in search nodes. Identical in every implementation. */
    private static final int SIPHON_SEARCH_BUDGET = 10_000;

    private StructuralCheck() {}

    /**
     * Result of structural deadlock check.
     */
    public sealed interface Result {
        /** Every minimal siphon contains a trap marked in the initial marking. */
        record NoPotentialDeadlock() implements Result {}

        /** A siphon was found that does not contain a marked trap. */
        record PotentialDeadlock(Set<Integer> siphon) implements Result {}

        /** Analysis could not decide: the net has too many places, or the siphon search exceeded its node budget. */
        record Inconclusive(String reason) implements Result {}
    }

    /**
     * Checks for potential deadlocks using siphon/trap analysis.
     *
     * @param flatNet the flattened net
     * @param initialMarking the initial marking
     * @return the check result; {@code Inconclusive} when the net is too large or the siphon search exceeds its budget
     */
    public static Result check(FlatNet flatNet, MarkingState initialMarking) {
        int P = flatNet.placeCount();

        if (P == 0) {
            return new Result.NoPotentialDeadlock();
        }

        // For large nets, skip (too expensive for complete enumeration)
        if (P > MAX_PLACES_FOR_SIPHON_ANALYSIS) {
            return new Result.Inconclusive("Net has " + P + " places, siphon enumeration skipped");
        }

        var siphons = findMinimalSiphons(flatNet, SIPHON_SEARCH_BUDGET);
        if (siphons == null) {
            return new Result.Inconclusive("siphon search exceeded " + SIPHON_SEARCH_BUDGET + " nodes");
        }

        // For each siphon, check if it contains a marked trap
        for (var siphon : siphons) {
            // Find the maximal trap within this siphon
            var trap = findMaximalTrapIn(flatNet, siphon);

            // Check if the trap is initially marked
            if (trap.isEmpty() || !isMarked(trap, flatNet, initialMarking)) {
                return new Result.PotentialDeadlock(siphon);
            }
        }

        return new Result.NoPotentialDeadlock();
    }

    /**
     * Finds <b>all</b> minimal siphons, or {@code null} when the search exceeds
     * {@code budget} nodes.
     *
     * <p>A siphon is a place set S such that every transition with an output in S has
     * at least one input in S. The search grows S from each start place; where a
     * producer into S has no input in S, it branches on each of that producer's inputs.
     * Every minimal siphon containing the start place is reached by the branch that
     * always picks an input inside it, so the search is complete. Committing to one
     * input instead (the first, or all of them at once) is not: it can miss exactly
     * the unmarked siphon that makes the net dead.
     */
    static List<Set<Integer>> findMinimalSiphons(FlatNet flatNet, int budget) {
        var found = new ArrayList<Set<Integer>>();
        var nodes = new int[] {0};
        for (int start = 0; start < flatNet.placeCount(); start++) {
            if (!growSiphon(flatNet, Set.of(start), found, nodes, budget)) {
                return null;
            }
        }
        var minimal = new ArrayList<Set<Integer>>();
        for (int i = 0; i < found.size(); i++) {
            var s = found.get(i);
            boolean isMinimal = true;
            for (int j = 0; j < found.size(); j++) {
                var other = found.get(j);
                if (j != i && other.size() <= s.size() && s.containsAll(other)) {
                    isMinimal = false;
                    break;
                }
            }
            if (isMinimal) {
                minimal.add(s);
            }
        }
        return minimal;
    }

    /** One node of the siphon search. Returns {@code false} when the budget is exhausted. */
    private static boolean growSiphon(
            FlatNet flatNet, Set<Integer> siphon, List<Set<Integer>> found, int[] nodes, int budget) {
        if (++nodes[0] > budget) {
            return false;
        }
        // A superset of a siphon already found cannot lead to a new minimal one.
        for (var f : found) {
            if (f.size() <= siphon.size() && siphon.containsAll(f)) {
                return true;
            }
        }

        FlatTransition violating = null;
        for (var ft : flatNet.transitions()) {
            boolean outputsToSiphon = false;
            boolean hasInputInSiphon = false;
            for (int p : siphon) {
                if (ft.postVector()[p] > 0) outputsToSiphon = true;
                if (ft.preVector()[p] > 0) hasInputInSiphon = true;
            }
            if (outputsToSiphon && !hasInputInSiphon) {
                violating = ft;
                break;
            }
        }

        if (violating == null) {
            found.add(siphon);
            return true;
        }
        // A producer with no inputs keeps any set holding its output marked: no siphon
        // on this branch. Otherwise branch on each input.
        for (int q = 0; q < flatNet.placeCount(); q++) {
            if (violating.preVector()[q] > 0) {
                var next = new TreeSet<>(siphon);
                next.add(q);
                if (!growSiphon(flatNet, Collections.unmodifiableSortedSet(next), found, nodes, budget)) {
                    return false;
                }
            }
        }
        return true;
    }

    /**
     * Finds the maximal trap within a given set of places.
     * Uses fixed-point: start with the full set and remove places that violate the trap condition.
     */
    static Set<Integer> findMaximalTrapIn(FlatNet flatNet, Set<Integer> places) {
        var trap = new TreeSet<>(places);

        boolean changed = true;
        while (changed) {
            changed = false;
            var toRemove = new ArrayList<Integer>();

            for (int p : trap) {
                // Check trap condition: every transition consuming from p must produce to some place in trap
                boolean satisfies = true;
                for (int t = 0; t < flatNet.transitionCount(); t++) {
                    var ft = flatNet.transitions().get(t);
                    if (ft.preVector()[p] > 0) {
                        // This transition consumes from p - check it outputs to some trap place
                        boolean outputsToTrap = false;
                        for (int q : trap) {
                            if (ft.postVector()[q] > 0) {
                                outputsToTrap = true;
                                break;
                            }
                        }
                        if (!outputsToTrap) {
                            satisfies = false;
                            break;
                        }
                    }
                }
                if (!satisfies) {
                    toRemove.add(p);
                }
            }

            if (!toRemove.isEmpty()) {
                trap.removeAll(toRemove);
                changed = true;
            }
        }

        return Set.copyOf(trap);
    }

    /**
     * Checks if a set of places has at least one token in the initial marking.
     */
    private static boolean isMarked(Set<Integer> placeIndices, FlatNet flatNet, MarkingState marking) {
        for (int idx : placeIndices) {
            var place = flatNet.places().get(idx);
            if (marking.tokens(place) > 0) {
                return true;
            }
        }
        return false;
    }
}
